---
output:
  reprex::reprex_document:
    venue: "gh"
    advertise: FALSE
    session_info: TRUE
    style: TRUE
    comment: "#;-)"
    tidyverse_quiet: FALSE
    std_out_err: TRUE
knit: reprex::reprex_render
---

```{r, echo = FALSE, warning= FALSE}
# Loading libraries and setting working directory
library(vcfR)
library(tidyverse)

setwd('~/cran/VCFs/')
```

```{r}
# Loading SNPs data
vcf <- read.vcfR('filtered.vcf', convertNA = T)
```

```{r}
# Extract genotypes
snps_num <- extract.gt(vcf, 
                       element = "GT",
                       IDtoRowNames  = F,
                       as.numeric = T,
                       convertNA = T,
                       return.alleles = F)
gt_info <- snps_num # Getting a backup

# How many NAs are for each sample?
na_info <- colSums(is.na(gt_info)) 
summary(na_info)

# Remove samples with more than 25% NAs
# 25% threshold: 
thres <- round(nrow(gt_info) * 0.25)

# Calculate number of NAs in each col
na_counts <- na_info %>% as.data.frame() %>% `colnames<-`('perc')

# Filter samples with high % (>75) of NAs
snps_num_df <- snps_num[na_counts$perc <= thres,]

dim(snps_num) - dim(snps_num_df) # How many samples were removed?
```

```{r}
# Remove invariable SNPs
# Calculate the SD for each marker across all the samples
ds_snps <- apply(snps_num_df, 1, sd, na.rm = T)
table(ds_snps == 0)

```



```{r}
snps <- snps_num_df[!(ds_snps == 0),] 
apply(snps, 1, sd, na.rm = T) %>% summary()

```
